src/util/renderer.ts

```diff
@@ -0,0 +1,352 @@
+	}
+
+	static bundle(device: GPUDevice, bundleElement: BundleElement, ): GPURenderBundle {
+		const b = bundleElement;
+		const encoder = device.createRenderBundleEncoder({
+			colorFormats: [Renderer.colorFormat],
+			depthStencilFormat: Renderer.depthFormat
+		});
+		encoder.setPipeline(b.pipeline);
+		for (let i = 0; i < b.vertexBuffers.length; i++) {
+			encoder.setVertexBuffer(i, b.vertexBuffers[i]);
+		}
+		for (let i = 0; i < b.bindGroups.length; i++) {
+			encoder.setBindGroup(i, b.bindGroups[i]);
+		}
+		if (b.drawCounts instanceof Array) {
+			encoder.draw(b.drawCounts[0], b.drawCounts[1] ?? 1, b.drawCounts[2] ?? 0, b.drawCounts[3] ?? 0);
+		} else {
+			encoder.draw(b.drawCounts.vertexCount, b.drawCounts.instanceCount ?? 1, b.drawCounts.firstVertex ?? 0, b.drawCounts.firstInstance ?? 0);
+		}
+		const bundle = encoder.finish();
+		Renderer._bundles.push(bundle);
+		return bundle;
+	}
+
+	static clearQueue() {
+		Renderer._queue = [];
+	}
+
+	static clearBundles() {
+		Renderer._bundles = [];
+	}
+
+	static async submitQueue(device?: GPUDevice) {
+		const commands: GPUCommandBuffer[] = [];
+		for (let i = 0; i < Renderer._queue.length; i++) {
+			const q = Renderer._queue[i];
+			if (device) {
+				commands.push(Renderer.render(q, false));
+			} else {
+				Renderer.render(q);
+			}
+		}
+		if (device && commands.length > 0) {
+			device.queue.submit(commands);
+		}
+	}
+
+	static async submitBundles(device: GPUDevice, renderPassDescriptor: GPURenderPassDescriptor) {
+		if (device && Renderer._bundles.length > 0) {
+			const commandEncoder = device.createCommandEncoder();
+			const passEncoder = commandEncoder.beginRenderPass(renderPassDescriptor);
+			passEncoder.executeBundles(Renderer._bundles);
+			passEncoder.end();
+			device.queue.submit([commandEncoder.finish()]);
+		}
+	}
+
+	static createRenderPipeline(
+		name: string,
+		device: GPUDevice,
+		shader: GPUShaderModule,
+		format: GPUTextureFormat,
+		buffers: Iterable<GPUVertexBufferLayout | null>,
+		bindGroupLayout?: GPUPipelineLayout
+	): GPURenderPipeline {
+		return device.createRenderPipeline({
+			label: name + ' Render Pipeline',
+			layout: bindGroupLayout ?? 'auto',
+			vertex: {
+				module: shader,
+				entryPoint: 'main_vertex',
+				buffers
+			},
+			fragment: {
+				module: shader,
+				entryPoint: 'main_fragment',
+				targets: [
+					{
+						format,
+						blend: {
+							alpha: {
+								srcFactor: 'one',
+								dstFactor: 'one-minus-src-alpha',
+								operation: 'add',
+							},
+							color: {
+								srcFactor: 'src-alpha',
+								dstFactor: 'one-minus-src-alpha',
+								operation: 'add',
+							},
+						} as GPUBlendState,
+					},
+				],
+			},
+			primitive: {
+				topology: 'triangle-list',
+			},
+			depthStencil: {
+				format: 'depth24plus',
+				depthCompare: 'less-equal',
+				depthWriteEnabled: true,
+			},
+		});
+	}
+
+	static createPipelineLayout(name: string, device: GPUDevice, bindGroupLayouts: GPUBindGroupLayout[]) {

```

Comment:
This was a mistake in my original implementation. We should not be creating a pipeline for each mark type, since that has major performance costs. We should instead try to create a single pipeline that uses [overrides](https://www.w3.org/TR/WGSL/#override-decls) to handle each mark type.